### PR TITLE
Feature SEO for OSF

### PR DESCRIPTION
**Pull request by kushG**
<h3> Purpose </h3>

OSF is currently using a lot of javascript to render html content. This makes applications faster but would be least searchable since most of the contents are generated dynamically & search engines can only see static contents. This asks for creating & saving the HTML content after Javascript is executed, also known as static snaphot. This PR creates static snapshot for public pages on OSF & returns these pages when Google's search engine(i.e. crawler) makes a request.

<h3> Changes </h3>

<strong> Theory </strong> : Basically, Google crawler makes a request to OSF only if OSF has some kind of indicator that crawler expects. This indicator could either be adding a '!'  in the url of a page (before hash fragments) or adding a meta tag(&lt;meta name="fragment" content="!">) to HTML for that page.  This PR uses the latter. Once, OSF is google crawlable every request to that OSF page would also get an additional request from Google crawler with a query paramater(say, ?_=123456). This request should then recieve static snapshot of that page as response.

This PR adds meta tag to base.mako which applies to all the pages on OSF. Whenever a request is made for a page, a celery background task is run which uses PhantomJS to render the javascript enabled HTML static snapshot of the page. This page content would then be cached on disk & retrieved when a Google crawler request is made.

Major changes in this PR would be a decorator added that runs the background tasks & a view function that handles's the request made by Google crawler.

<h3> Side Effects </h3>
Any error caused by the new decorator that has ben added may result in failure of the corresponding view function.

### CODE REVIEW

**Comment by lbanner** [comment]
:+1:

File: website/static/js/phantomServer/app.js [hunk 1]
```diff
@@ -0,0 +1,48 @@
+var express = require('express');
+var phantom = require('phantom');
+var app = express();
+
+app.get('/', function (req, res) {
+
+    var url = req.query.url;
+    console.log(url);
+    console.log(req.query);
```
Comment (sloria) [inline website/static/js/phantomServer/app.js]: Could make logging more useful here, e.g. `Opening "http://osf.io/oaiwje/" with query parameters {foo: "bar"}`

File: website/static/js/phantomServer/routes/index.js [hunk 2]
```diff
@@ -0,0 +1,9 @@
+var express = require('express');
+var router = express.Router();
+
+/* GET home page. */
+router.get('/', function(req, res, next) {
```
Comment (sloria) [inline website/static/js/phantomServer/routes/index.js]: Is this route necessary?

File: website/static/js/phantomServer/app.js [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+    });
+
+});
+
+var server = app.listen(3000, function () {
+
+  var host = server.address().address;
+  var port = server.address().port;
+
+  console.log('Example app listening at http://%s:%s', host, port);
```
Comment (sloria) [inline website/static/js/phantomServer/app.js]: Change `Example app` to something like `OSF PhantomJS app`

File: website/static_snapshot/__init__.py [hunk 4]
```diff
@@ -0,0 +1,4 @@
+SNAPSHOT_PAGE = {
```
Comment (sloria) [inline website/static_snapshot/__init__.py]: This doesn't seem to be used anywhere.

File: website/views.py [hunk 5]
```diff
@@ -219,8 +222,19 @@ def get_dashboard_nodes(auth):
     return _render_nodes(response_nodes, auth)
 
 
+def homepage_snapshot():
```
Comment (sloria) [inline website/views.py]: Why is this necessary?

**Comment by sloria** [comment]
Please add tests for the new decorator, utilities, etc.

File: package.json [hunk 6]
```diff
@@ -17,7 +17,7 @@
     "diff": "~1.2.2",
     "dropzone": "git://github.com/sloria/dropzone.git#accept-directory",
     "exports-loader": "^0.6.2",
-    "express": "~4.10.0",
+    "express": "^4.10.8",
```
Comment (sloria) [inline package.json]: Better to pin to an exact version if possible rather than using `^`. If not possible due to a conflicting dependency, use `~`. See: http://stackoverflow.com/a/22345808

File: website/static/js/phantomServer/package.json [hunk 7]
```diff
@@ -0,0 +1,17 @@
+{
+  "name": "phantomServer",
+  "version": "0.0.0",
+  "private": true,
+  "scripts": {
+    "start": "node ./bin/www"
```
Comment (sloria) [inline website/static/js/phantomServer/package.json]: This seems incorrect. Shouldn't it be `node ./app.js`? Though this probably isn't necessary, given that we already have an invoke task.

File: website/views.py (hunk 5 above)
Comment (kushG) [inline website/views.py]: This is required to make crawler call the homepage(i.e. osf.io). Usually, crawlwer calls are made when a javascript makes api calls to OSF, a corresponding crawler call is made. Since there were no API calls made when homepage was called, I've added this API view function.

File: website/static/js/phantomServer/README.md [hunk 8]
```diff
@@ -0,0 +1,3 @@
+1. In website/settings/local.py turn USE_CELERY to True
+2. Run 'invoke phantom' to start phantom app using Express Server.
+3. Run 'invoke rabbitmq' & 'invoke celery_worker' for runing celery asynchronous tasks.
```
Comment (lyndsysimon) [inline website/static/js/phantomServer/README.md]: Minor typo: "runing"

**Comment by icereval** [comment]
Closing this PR as https://github.com/CenterForOpenScience/prerender has been deployed
